python/src/exactextract/writer.py

```diff
@@ -129,6 +129,169 @@ def features(self):
             return pd.DataFrame(self.fields, copy=False)
 
 
+class QGISWriter(Writer):
+    """Creates QGSVectorLayer"""
+
+    def __init__(self, *, srs_wkt=None):
+        super().__init__()
+        from qgis.core import QgsCoordinateReferenceSystem
+
+        self.fields = []
+        self.feats = []
+        self.geom_type = "unknown"
+        self.fields_set = False
+
+        self.crs = QgsCoordinateReferenceSystem()
+        self.crs.createFromWkt(srs_wkt)
+
+    def add_operation(self, op):
+        self.fields.append(op.name)
+
+    def add_column(self, col_name):
+        self.fields.append(col_name)
+
+    def write(self, feature):
+        import json
+
+        from qgis.core import QgsFeature, QgsJsonUtils
+
+        f = JSONFeature()
+        feature.copy_to(f)
+
+        if not self.fields_set:
+            # test every property vs fields type and set correct type
+            self.fields = self._set_fields_types(self.fields, f)
+            self.fields = self._convert_to_QgsFields(self.fields)
+            self.geom_type = self._get_geometry_type(f)
+            self.fields_set = True
+
+        # create a QgsFeature and set fields
+        fet = QgsFeature()
+        fet.setFields(self.fields)
+        # populate QgsFeature with values
+        for field_name, value in f.feature["properties"].items():
+            field_type = self.fields.at(
+                self.fields.indexFromName(field_name)
+            ).typeName()
+            if field_type == "Array":
+                value = str(value)
+            fet.setAttribute(field_name, value)
+        # set id if present
+        if "id" in f.feature:
+            fet.setAttribute("id", f.feature["id"])
+        # set geometry based on JSON geometry
+        fet.setGeometry(
+            QgsJsonUtils.geometryFromGeoJson(json.dumps(f.feature["geometry"]))
+        )
+
+        self.feats.append(fet)
```

Comment:
Can the `write` method directly write the feature to a member`QgsVectorLayer`? (This is how `GDALWriter` works)

Reply:
I think it's possible. My only concern there is the speed of writing it to the `QgsVectorLayer` in `write` compared to the current solution. I will benchmark this.

Reply:
@dbaston Do you know if I can access the `feature` inside `__init__` or its geometry type? `QgsVectorLayer` requires knowledge about features geometry type that will be put in the layer during its definition in the `memory` layer I am using there.
I might use the current "workaround" (`if not self.fields_set:`) and define the `QgsVectorLayer` when the first feature is going to be written but it doesn't seem to be the elegant solution.

Reply:
> define the QgsVectorLayer when the first feature is going to be written

This sound like the simplest way to me.